File: sdk/src/main/java/com/libreshockwave/format/AfterburnerReader.java
package com.libreshockwave.format;

import com.libreshockwave.io.BinaryReader;

import java.io.IOException;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Reader for Afterburner-compressed Director files (.dcr).
 * Handles the FGDM/FGDC container format with zlib compression.
 */
public class AfterburnerReader {

    private final BinaryReader reader;
    private final ByteOrder byteOrder;

    // Parsed metadata
    private int directorVersion;
    private int imapVersion;
    private String versionString;

    // Compression types from Fcdr chunk
    private final List<MoaID> compressionTypes = new ArrayList<>();

    // Resource map from ABMP chunk
    private final Map<Integer, ChunkInfo> chunkInfoMap = new LinkedHashMap<>();

    // ILS (Initial Load Segment) data
    private byte[] ilsData;
    private int ilsBodyOffset;

    // Cached chunk data extracted from ILS
    private final Map<Integer, byte[]> cachedChunkData = new HashMap<>();

    public AfterburnerReader(BinaryReader reader, ByteOrder byteOrder) {
        this.reader = reader;
        this.byteOrder = byteOrder;
    }

    /**
     * Parse the Afterburner map and prepare for chunk extraction.
     */
    public void parse() throws IOException {
        // Read Fver (File Version) chunk
        readFverChunk();

        // Read Fcdr (Compression Directory) chunk
        readFcdrChunk();

        // Read ABMP (Afterburner Map) chunk
        readAbmpChunk();

        // Read FGEI/ILS (Initial Load Segment) chunk
        readIlsChunk();
    }

    /**
     * Read the Fver (File Version) chunk.
     */
    private void readFverChunk() throws IOException {
        String fourcc = readFourCCOrdered();
        if (!fourcc.equals("Fver")) {
            throw new IOException("Expected Fver chunk, got: " + fourcc);
        }

        int length = readVarInt();
        int endPos = reader.getPosition() + length;

        // Sanity check - endPos should be within file bounds
        if (endPos > reader.length()) {
            throw new IOException("Fver chunk length " + length + " extends past end of file (endPos=" + endPos + ", fileLen=" + reader.length() + ")");
        }

        // Read version fields
        imapVersion = readVarInt();
        directorVersion = readVarInt();

        // Version string (optional, v0x501+)
        if (reader.getPosition() < endPos && reader.bytesLeft() > 0) {
            int strLen = readVarInt();
            int maxStrLen = endPos - reader.getPosition();
            if (strLen > 0 && strLen <= maxStrLen && strLen < 10000) {
                byte[] strBytes = reader.readBytes(strLen);
                versionString = new String(strBytes).trim();
            }
        }

        // Skip any remaining data (but don't seek past file end)
        if (endPos <= reader.length()) {
            reader.seek(endPos);
        }
    }

    /**
     * Read the Fcdr (Compression Directory) chunk.
     */
    private void readFcdrChunk() throws IOException {
        String fourcc = readFourCCOrdered();
        if (!fourcc.equals("Fcdr")) {
            throw new IOException("Expected Fcdr chunk, got: " + fourcc);
        }

        int compressedLength = readVarInt();
        byte[] compressedData = reader.readBytes(compressedLength);
        byte[] data = reader.decompressZlib(compressedData);

        BinaryReader fcdrReader = new BinaryReader(data, byteOrder);

        // Read compression type count
        int typeCount = fcdrReader.readUnsignedShort();

        // Read MoaID compression types
        for (int i = 0; i < typeCount; i++) {
            MoaID moaId = MoaID.read(fcdrReader);
            compressionTypes.add(moaId);
        }

        // Read compression type description strings (null-terminated)
        for (int i = 0; i < typeCount; i++) {
            int b;
            while ((b = fcdrReader.readUnsignedByte()) != 0) {
                // Description string is informational only, not stored
            }
        }
    }

    /**
     * Read the ABMP (Afterburner Map) chunk.
     */
    private void readAbmpChunk() throws IOException {
        String fourcc = readFourCCOrdered();
        if (!fourcc.equals("ABMP")) {
            throw new IOException("Expected ABMP chunk, got: " + fourcc);
        }

        int chunkLength = readVarInt();
        int chunkEndPos = reader.getPosition() + chunkLength;

        // Read header fields: unknown VarInt and uncompressed length VarInt
        readVarInt(); // unk
        readVarInt(); // uncompLength

        // Read remaining bytes as compressed data
        int compressedLength = chunkEndPos - reader.getPosition();
        byte[] compressedData = reader.readBytes(compressedLength);
        byte[] data = reader.decompressZlib(compressedData);

        BinaryReader abmpReader = new BinaryReader(data, byteOrder);

        // Unknown fields
        readVarInt(abmpReader); // unk1
        readVarInt(abmpReader); // unk2

        // Resource count
        int resCount = readVarInt(abmpReader);

        // Read resource entries
        // Note: offset can be -1 (0xFFFFFFFF as VarInt), meaning "accumulate from previous"
        // The first entry (ILS) describes the ILS blob itself and uses offset=0
        // Subsequent entries are chunks within the DECOMPRESSED ILS
        int runningOffset = 0;
        for (int i = 0; i < resCount; i++) {
            int resId = readVarInt(abmpReader);
            int rawOffset = readVarInt(abmpReader);
            int compSize = readVarInt(abmpReader);
            int uncompSize = readVarInt(abmpReader);
            int compressionTypeIndex = readVarInt(abmpReader);
            // Tag is stored as a u32 in the file's byte order. In LE files, tags are
            // byte-swapped (e.g., "STXT" stored as 54 58 54 53). readI32() with file's
            // endian produces the correct integer, and fourCCToString converts it back.
            int tagInt = abmpReader.readI32();
            String tag = BinaryReader.fourCCToString(tagInt);

            // Calculate actual offset
            int offset;
            if (rawOffset == 0 && i == 0 && tag.trim().equals("ILS")) {
                // First entry with offset=0 and tag ILS is the ILS descriptor
                // It doesn't have a real offset within the decompressed data
                offset = 0;  // Special: this entry is meta-info, not extractable
            } else if (rawOffset == -1) {
                // -1 means "accumulate from previous"
                offset = runningOffset;
                runningOffset = offset + compSize;
            } else {
                offset = rawOffset;
                runningOffset = offset + compSize;
            }

            MoaID compressionType = compressionTypeIndex < compressionTypes.size()
                ? compressionTypes.get(compressionTypeIndex)
                : MoaID.NULL_COMPRESSION;

            ChunkInfo info = new ChunkInfo(resId, tag, offset, compSize, uncompSize, compressionType);
            chunkInfoMap.put(resId, info);
        }
    }

    /**
     * Read the FGEI/ILS (Initial Load Segment) chunk.
     */
    private void readIlsChunk() throws IOException {
        String fourcc = readFourCCOrdered();
        if (!fourcc.equals("FGEI")) {
            throw new IOException("Expected FGEI chunk, got: " + fourcc);
        }

        // The ILS entry (ID=2) in ABMP tells us the compressed/uncompressed sizes
        ChunkInfo ilsInfo = chunkInfoMap.get(2);
        if (ilsInfo == null) {
            throw new IOException("No ILS entry (ID=2) found in ABMP");
        }

        // Read the unknown field (usually 0)
        readVarInt();

        // Store the position where ILS body starts
        ilsBodyOffset = reader.getPosition();

        // Read and decompress the ILS data using size from ABMP
        byte[] compressedData = reader.readBytes(ilsInfo.compressedSize());
        ilsData = reader.decompressZlib(compressedData);

        // Parse the ILS stream: it contains (resId, data) pairs
        // where data length comes from ABMP chunkInfo
        BinaryReader ilsReader = new BinaryReader(ilsData, byteOrder);
        while (!ilsReader.eof() && ilsReader.bytesLeft() > 0) {
            int resId = readVarInt(ilsReader);
            ChunkInfo info = chunkInfoMap.get(resId);
            if (info == null) {
                break;
            }

            // Read the chunk data using length from ABMP
            int dataLen = info.compressedSize();
            if (ilsReader.bytesLeft() < dataLen) {
                break;
            }

            byte[] chunkData = ilsReader.readBytes(dataLen);
            cachedChunkData.put(resId, chunkData);
        }
    }

    /**
     * Get chunk data by resource ID.
     */
    public byte[] getChunkData(int resourceId) throws IOException {
        // Skip the ILS meta entry (ID=2) - it describes the ILS blob, not a chunk
        if (resourceId == 2) {
            ChunkInfo info = chunkInfoMap.get(resourceId);
            if (info != null && info.fourCC().trim().equals("ILS")) {
                return ilsData;  // Return the decompressed ILS data itself
            }
        }

        // Check if already cached (from ILS parsing)
        if (cachedChunkData.containsKey(resourceId)) {
            byte[] chunkData = cachedChunkData.get(resourceId);

            // Decompress if needed (check if sizes differ)
            ChunkInfo info = chunkInfoMap.get(resourceId);
            if (info != null && info.compressedSize() != info.uncompressedSize()) {
                if (info.isZlibCompressed()) {
                    chunkData = reader.decompressZlib(chunkData);
                }
            }
            return chunkData;
        }

        ChunkInfo info = chunkInfoMap.get(resourceId);
        if (info == null) {
            throw new IOException("Unknown resource ID: " + resourceId);
        }

        // For non-ILS chunks, read from the remaining file data after ILS
        // The offset in ABMP is relative to ilsBodyOffset
        int fileOffset = ilsBodyOffset + info.offset();
        if (fileOffset >= reader.length()) {
            throw new IOException("Chunk " + resourceId + " (" + info.fourCC() +
                ") offset " + fileOffset + " exceeds file length " + reader.length());
        }

        // Save current position and seek to chunk
        int savedPos = reader.getPosition();
        reader.seek(fileOffset);

        byte[] chunkData = reader.readBytes(info.compressedSize());

        // Restore position
        reader.seek(savedPos);

        // Decompress if needed
        if (info.compressedSize() != info.uncompressedSize() && info.isZlibCompressed()) {
            chunkData = reader.decompressZlib(chunkData);
        }

        // Cache for future use
        cachedChunkData.put(resourceId, chunkData);
        return chunkData;
    }

    /**
     * Get chunk data by FourCC type.
     * Returns the first matching chunk.
     */
    public byte[] getChunkDataByType(String fourCC) throws IOException {
        for (ChunkInfo info : chunkInfoMap.values()) {
            if (info.fourCC().equals(fourCC)) {
                return getChunkData(info.resourceId());
            }
        }
        throw new IOException("Chunk type not found: " + fourCC);
    }

    /**
     * Get all chunk infos.
     */
    public Collection<ChunkInfo> getChunkInfos() {
        return Collections.unmodifiableCollection(chunkInfoMap.values());
    }

    /**
     * Get chunk info by resource ID.
     */
    public ChunkInfo getChunkInfo(int resourceId) {
        return chunkInfoMap.get(resourceId);
    }

    /**
     * Get all chunks of a specific type.
     */
    public List<ChunkInfo> getChunksByType(String fourCC) {
        List<ChunkInfo> result = new ArrayList<>();
        for (ChunkInfo info : chunkInfoMap.values()) {
            if (info.fourCC().equals(fourCC)) {
                result.add(info);
            }
        }
        return result;
    }

    /**
     * Read a FourCC string respecting the byte order.
     * For little-endian files, the FourCC is stored reversed.
     */
    private String readFourCCOrdered() {
        return readFourCCOrdered(reader);
    }

    /**
     * Read a FourCC string from a specific reader, respecting the byte order.
     */
    private String readFourCCOrdered(BinaryReader r) {
        byte[] bytes = r.readBytes(4);
        if (byteOrder == ByteOrder.LITTLE_ENDIAN) {
            // Reverse the bytes for little-endian
            byte temp = bytes[0];
            bytes[0] = bytes[3];
            bytes[3] = temp;
            temp = bytes[1];
            bytes[1] = bytes[2];
            bytes[2] = temp;
        }
        return new String(bytes, java.nio.charset.StandardCharsets.US_ASCII);
    }

    /**
     * Read a variable-length integer (VarInt).
     * Used extensively in Afterburner format.
     */
    private int readVarInt() throws IOException {
        return readVarInt(reader);
    }

    /**
     * Read a variable-length integer using MSB-first encoding.
     * High bit set means more bytes follow.
     */
    private int readVarInt(BinaryReader r) throws IOException {
        int value = 0;
        int b;
        do {
            b = r.readUnsignedByte();
            value = (value << 7) | (b & 0x7F);
        } while ((b & 0x80) != 0);
        return value;
    }

    /**
     * Calculate size of a VarInt value.
     */
    private int getVarIntSize(int value) {
        int size = 1;
        while (value >= 0x80) {
            size++;
            value >>= 7;
        }
        return size;
    }

    // Getters

    public int getDirectorVersion() {
        return directorVersion;
    }

    public int getImapVersion() {
        return imapVersion;
    }

    public String getVersionString() {
        return versionString;
    }

    public List<MoaID> getCompressionTypes() {
        return Collections.unmodifiableList(compressionTypes);
    }

    public int getChunkCount() {
        return chunkInfoMap.size();
    }
}

File: cpp/src/format/AfterburnerReader.cpp
#include "libreshockwave/format/AfterburnerReader.hpp"

#include <algorithm>
#include <stdexcept>
#include <utility>

namespace libreshockwave::format {
namespace {

std::string trimChunkTag(std::string tag) {
    while (!tag.empty() && (tag.back() == ' ' || tag.back() == '\0')) {
        tag.pop_back();
    }
    return tag;
}

} // namespace

AfterburnerReader::AfterburnerReader(io::BinaryReader reader, io::ByteOrder byteOrder)
    : reader_(std::move(reader)), byteOrder_(byteOrder) {
    reader_.setOrder(byteOrder_);
}

void AfterburnerReader::parse() {
    readFverChunk();
    readFcdrChunk();
    readAbmpChunk();
    readIlsChunk();
}

std::optional<std::vector<std::uint8_t>> AfterburnerReader::getChunkData(int resourceId) {
    if (resourceId == 2) {
        const auto* info = getChunkInfo(resourceId);
        if (info && trimChunkTag(info->fourCC) == "ILS") {
            return ilsData_;
        }
    }

    if (const auto cached = cachedChunkData_.find(resourceId); cached != cachedChunkData_.end()) {
        auto chunkData = cached->second;
        const auto* info = getChunkInfo(resourceId);
        if (info && info->compressedSize != info->uncompressedSize && info->isZlibCompressed()) {
            chunkData = io::BinaryReader::decompressZlib(chunkData);
        }
        return chunkData;
    }

    const auto* info = getChunkInfo(resourceId);
    if (!info) {
        return std::nullopt;
    }

    const int fileOffset = ilsBodyOffset_ + info->offset;
    if (fileOffset < 0 || static_cast<std::size_t>(fileOffset) >= reader_.length() ||
        info->compressedSize < 0 ||
        static_cast<std::size_t>(info->compressedSize) > reader_.length() - static_cast<std::size_t>(fileOffset)) {
        return std::nullopt;
    }

    const auto savedPosition = reader_.position();
    reader_.seek(static_cast<std::size_t>(fileOffset));
    auto chunkData = reader_.readBytes(static_cast<std::size_t>(info->compressedSize));
    reader_.seek(savedPosition);

    if (info->compressedSize != info->uncompressedSize && info->isZlibCompressed()) {
        chunkData = io::BinaryReader::decompressZlib(chunkData);
    }

    cachedChunkData_[resourceId] = chunkData;
    return chunkData;
}

std::optional<std::vector<std::uint8_t>> AfterburnerReader::getChunkDataByType(const std::string& fourCC) {
    for (const auto& [resourceId, info] : chunkInfoMap_) {
        if (info.fourCC == fourCC) {
            return getChunkData(resourceId);
        }
    }
    return std::nullopt;
}

std::vector<ChunkInfo> AfterburnerReader::getChunksByType(const std::string& fourCC) const {
    std::vector<ChunkInfo> result;
    for (const auto& [resourceId, info] : chunkInfoMap_) {
        if (info.fourCC == fourCC) {
            result.push_back(info);
        }
    }
    return result;
}

void AfterburnerReader::clearCachedData() {
    cachedChunkData_.clear();
}

int AfterburnerReader::directorVersion() const { return directorVersion_; }
int AfterburnerReader::imapVersion() const { return imapVersion_; }
const std::string& AfterburnerReader::versionString() const { return versionString_; }
const std::vector<MoaID>& AfterburnerReader::compressionTypes() const { return compressionTypes_; }
int AfterburnerReader::chunkCount() const { return static_cast<int>(chunkInfoMap_.size()); }

std::vector<ChunkInfo> AfterburnerReader::chunkInfos() const {
    std::vector<ChunkInfo> result;
    result.reserve(chunkInfoMap_.size());
    for (const auto& [resourceId, info] : chunkInfoMap_) {
        result.push_back(info);
    }
    return result;
}

const ChunkInfo* AfterburnerReader::getChunkInfo(int resourceId) const {
    if (const auto found = chunkInfoMap_.find(resourceId); found != chunkInfoMap_.end()) {
        return &found->second;
    }
    return nullptr;
}

std::string AfterburnerReader::readFourCCOrdered() {
    return readFourCCOrdered(reader_);
}

std::string AfterburnerReader::readFourCCOrdered(io::BinaryReader& reader) const {
    auto bytes = reader.readBytes(4);
    if (byteOrder_ == io::ByteOrder::LittleEndian) {
        std::reverse(bytes.begin(), bytes.end());
    }
    return std::string(bytes.begin(), bytes.end());
}

int AfterburnerReader::readVarInt() {
    return readVarInt(reader_);
}

int AfterburnerReader::readVarInt(io::BinaryReader& reader) const {
    int value = 0;
    int byte = 0;
    do {
        byte = reader.readU8();
        value = (value << 7) | (byte & 0x7F);
    } while ((byte & 0x80) != 0);
    return value;
}

void AfterburnerReader::readFverChunk() {
    const auto fourCC = readFourCCOrdered();
    if (fourCC != "Fver") {
        throw std::runtime_error("Expected Fver chunk, got: " + fourCC);
    }

    const int length = readVarInt();
    const auto endPosition = reader_.position() + static_cast<std::size_t>(length);
    if (endPosition > reader_.length()) {
        throw std::runtime_error("Fver chunk extends past end of data");
    }

    imapVersion_ = readVarInt();
    directorVersion_ = readVarInt();

    if (reader_.position() < endPosition && reader_.bytesLeft() > 0) {
        const int stringLength = readVarInt();
        const auto maxStringLength = endPosition - reader_.position();
        if (stringLength > 0 && static_cast<std::size_t>(stringLength) <= maxStringLength && stringLength < 10000) {
            auto bytes = reader_.readBytes(static_cast<std::size_t>(stringLength));
            versionString_ = std::string(bytes.begin(), bytes.end());
            while (!versionString_.empty() && (versionString_.back() == ' ' || versionString_.back() == '\0')) {
                versionString_.pop_back();
            }
        }
    }

    reader_.seek(endPosition);
}

void AfterburnerReader::readFcdrChunk() {
    const auto fourCC = readFourCCOrdered();
    if (fourCC != "Fcdr") {
        throw std::runtime_error("Expected Fcdr chunk, got: " + fourCC);
    }

    const int compressedLength = readVarInt();
    auto data = io::BinaryReader::decompressZlib(reader_.readBytes(static_cast<std::size_t>(compressedLength)));
    io::BinaryReader fcdrReader(data, byteOrder_);

    const int typeCount = fcdrReader.readU16();
    compressionTypes_.reserve(static_cast<std::size_t>(typeCount));
    for (int index = 0; index < typeCount; ++index) {
        compressionTypes_.push_back(MoaID::read(fcdrReader));
    }

    for (int index = 0; index < typeCount && !fcdrReader.eof(); ++index) {
        while (!fcdrReader.eof() && fcdrReader.readU8() != 0) {
        }
    }
}

void AfterburnerReader::readAbmpChunk() {
    const auto fourCC = readFourCCOrdered();
    if (fourCC != "ABMP") {
        throw std::runtime_error("Expected ABMP chunk, got: " + fourCC);
    }

    const int chunkLength = readVarInt();
    const auto chunkEndPosition = reader_.position() + static_cast<std::size_t>(chunkLength);
    (void)readVarInt();
    (void)readVarInt();

    if (chunkEndPosition > reader_.length() || chunkEndPosition < reader_.position()) {
        throw std::runtime_error("ABMP chunk extends past end of data");
    }

    auto data = io::BinaryReader::decompressZlib(reader_.readBytes(chunkEndPosition - reader_.position()));
    io::BinaryReader abmpReader(data, byteOrder_);

    (void)readVarInt(abmpReader);
    (void)readVarInt(abmpReader);
    const int resourceCount = readVarInt(abmpReader);

    int runningOffset = 0;
    for (int index = 0; index < resourceCount; ++index) {
        const int resourceId = readVarInt(abmpReader);
        const int rawOffset = readVarInt(abmpReader);
        const int compressedSize = readVarInt(abmpReader);
        const int uncompressedSize = readVarInt(abmpReader);
        const int compressionTypeIndex = readVarInt(abmpReader);
        const auto tagInt = abmpReader.readU32();
        const auto tag = io::BinaryReader::fourCCToString(tagInt);

        int offset = rawOffset;
        if (rawOffset == 0 && index == 0 && trimChunkTag(tag) == "ILS") {
            offset = 0;
        } else {
            offset = rawOffset;
            runningOffset = offset + compressedSize;
        }
        (void)runningOffset;

        const auto compressionType = compressionTypeIndex >= 0 &&
                                     static_cast<std::size_t>(compressionTypeIndex) < compressionTypes_.size()
            ? compressionTypes_[static_cast<std::size_t>(compressionTypeIndex)]
            : MoaID::NULL_COMPRESSION;
        chunkInfoMap_.insert_or_assign(resourceId,
                                       ChunkInfo{resourceId, tag, offset, compressedSize, uncompressedSize, compressionType});
    }
}

void AfterburnerReader::readIlsChunk() {
    const auto fourCC = readFourCCOrdered();
    if (fourCC != "FGEI") {
        throw std::runtime_error("Expected FGEI chunk, got: " + fourCC);
    }

    const auto* ilsInfo = getChunkInfo(2);
    if (!ilsInfo) {
        throw std::runtime_error("No ILS entry found in ABMP");
    }

    (void)readVarInt();
    ilsBodyOffset_ = static_cast<int>(reader_.position());
    ilsData_ = io::BinaryReader::decompressZlib(reader_.readBytes(static_cast<std::size_t>(ilsInfo->compressedSize)));

    io::BinaryReader ilsReader(ilsData_, byteOrder_);
    while (!ilsReader.eof() && ilsReader.bytesLeft() > 0) {
        const int resourceId = readVarInt(ilsReader);
        const auto* info = getChunkInfo(resourceId);
        if (!info) {
            break;
        }

        const int dataLength = info->compressedSize;
        if (dataLength < 0) {
            break;
        }
        if (ilsReader.bytesLeft() < static_cast<std::size_t>(dataLength)) {
            cachedChunkData_[resourceId] = ilsReader.readBytes(ilsReader.bytesLeft());
            break;
        }

        cachedChunkData_[resourceId] = ilsReader.readBytes(static_cast<std::size_t>(dataLength));
    }
}

} // namespace libreshockwave::format
